### PR TITLE
Add unit tests to Web App

### PR DESCRIPTION
**Pull request by isabarros**
[WIP]
I've created the PR already just to register some findings and to discuss the implementation. Please check the comments.
Here is the checklist with the tests needed from the issue https://github.com/berinhard/pyp5js/issues/81:

**Route /**

- [x] Ensure the sketches are being listed with valid links
- [x] On GET should render index.html form (suggestion)

**Route /new-sketch/**

- [x] On GET should render the new sketch form
- [x] POST without sketch_name should render the form again with an error message
- [x] POST with an existing sketch_name should render the form again with an error message
- [x] POST with a valid sketch_name should create the new sketch directory and files and render success page

**Route /sketch/<sketch_name>**

- [x] 404 if sketch does not exist
- [x] Should return the sketch's index.html if it exists
- [x] Should add the sketch's static files when the route has an additional path, such as /sketch/foo/static/style.css
- [x] Set Content-Type=application/javascript header for the javascript files;
- [x] 404 if the static file does not exist

### CODE REVIEW

File: pyp5js/http/web_app.py [hunk 1]
```diff
@@ -14,6 +14,8 @@
 @app.route("/")
 def sketches_list_view():
     sketches = []
+    if not SKETCHBOOK_DIR.exists():
+        SKETCHBOOK_DIR.mkdir()
```
Comment (isabarros) [inline pyp5js/http/web_app.py]: The first error that I faced while running the tests was the directory was not created, so I had to copy this lines from the config.py file into the `sketches_list_view()`. Need to check if this is a good solution and if following this approach we want to remove those lines from the config.py file.

File: pytest.ini [hunk 2]
```diff
@@ -1,3 +1,3 @@
 [pytest]
 env =
-    SKETCHBOOK_DIR=tests-skecthbook
+    SKETCHBOOK_DIR=tests-sketchbook
```
Comment (isabarros) [inline pytest.ini]: I have fixed this typo, but after rebasing my branch noticed that you did that as well :P

File: dev-requirements.txt [hunk 3]
```diff
@@ -5,3 +5,5 @@ ipython==7.8.0
 pytest==4.5.0
 tox==3.10.0
 pytest-env==0.6.2
+Flask-Testing==0.7.1
+blinker==1.4
```
[the remark this answers is not in the document]
Comment (berinhard) [reply dev-requirements.txt]: This is ok for me :+1:

File: pytest.ini (hunk 2 above)
Comment (berinhard) [reply pytest.ini]: No worries...

File: pyp5js/http/web_app.py (hunk 1 above)
Comment (berinhard) [reply pyp5js/http/web_app.py]: A suggestion here is to create a `pytest.fixture` to always create this directory and avoid doing this check in the view and directory creation inside the view. This error you've had can be somehow related to an import order in the tests or stuff like that. Because when running pyp5js "for real" this directory will always be created given the import of the `pyp5js.config` module in the very first entry point of the lib.

My suggestion is for you to use the [Flask-Testing TestCase methods](https://github.com/berinhard/pyp5js/blob/develop/pyp5js/tests/test_commands.py#L13) `setUp` and `tearDown`  to create a base test class for pyp5js web app. Something like:

```python
class Pyp5jsWebTestCase(TestCase):

    def create_app(self):
        app = web_app
        app.config['TESTING'] = True
        return app    

    def setUp(self):
        if not SKETCHBOOK_DIR.exists():
            SKETCHBOOK_DIR.mkdir()

    def tearDown(self):
        if SKETCHBOOK_DIR.exists():
            shutil.rmtree(SKETCHBOOK_DIR)
```

File: pyp5js/tests/http/test_web_app.py [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+        app.config['TESTING'] = True
+        return app
+
+    def test_get_home_renders_index_template(self):
+        self.client.get('/')
+        self.assert_template_used('index.html')
+
+    def test_get_new_sketch_renders_new_sketch_form_template(self):
+        self.client.get('/new-sketch/')
+        self.assert_template_used('new_sketch_form.html')
```
Comment (berinhard) [inline pyp5js/tests/http/test_web_app.py]: I hope I'm not being to picky, but I think a better organization for the tests would be to have 1 test class per route. It's something really helpful when it comes to preparing data via `setUp` or as url routing. It also improves the readability of the tests as well. What do you think about this type of structure:

```python
class IndexViewTests(Pyp5jsWebTestCase):
    route = '/'

    def test_get_home_renders_index_template(self):
        self.client.get(self.route)
        self.assert_template_used('index.html')


class NewSketchViewTests(Pyp5jsWebTestCase):
    route = '/new-sketch/'

    def test_get_new_sketch_renders_new_sketch_form_template(self):
        self.client.get(self.route)
        self.assert_template_used('new_sketch_form.html')
```

**Comment by berinhard** [comment]
The code looks great so far! I left a few suggestions on the tests organizations though. Thank you so much for helping me with this!

File: pyp5js/tests/http/test_web_app.py (hunk 4 above)
Comment (isabarros) [reply pyp5js/tests/http/test_web_app.py]: @berinhard Sounds good! I haven't really thought about splitting into different test classes but it makes sense grouping the scenarios by route. Will do that!
Some questions:
1. Do you think that will be good to have them in different files (one per class) or is it ok having them all inside `test_web_app.py`? Taking a look on the other tests I saw that there is usually one class per file and some fixtures, but I really don't have strong feelings in this case. Either way is fine for me.
2. For the parent class `Pyp5jsWebTestCase` the idea here is to have a base class for the web app route tests with the boilerplate for setting up the Flask app (the method `def create_app(self):` would be part of this base class for instance)?

File: pyp5js/http/web_app.py (hunk 1 above)
Comment (isabarros) [reply pyp5js/http/web_app.py]: I think the problem with having this condition in the `setUp` and `tearDown` methods is that we aren't really testing the creation of the dir, we are assuming that it is granted and therefore the part of the code in which it exists (the `config.py` file) stays uncovered. So even if  someone erases those lines from the `config.py` file, all the tests will still be passing.

Comment (isabarros) [reply pyp5js/http/web_app.py]: But at the same time I still do not think that the responsibility of creating the dir should be given to `sketches_list_view`, so I am really hesitating on where it should be placed

Comment (berinhard) [reply pyp5js/http/web_app.py]: I think you're totally correct about your points and your doubt exposes a flaw in `pyp5js`'s architecture. The `config` should be a lazy-load object instead of a python module as now. To unblock you from that, I've created #94 to address a refactor on how pyp5js manages configuration.

So, let's go with the `setUp`/`tearDown` approach for now so we can have a better test coverage that'll help with the config refactoring. Is that ok for you?

File: pyp5js/tests/http/test_web_app.py (hunk 4 above)
Comment (berinhard) [reply pyp5js/tests/http/test_web_app.py]: 1 - I don't think we need to bring this level of complexity right now. I mean, this is probably what we'll want to have in the future, but since now we only have a single `web_app.py` file, I think it's ok for the tests to reflect that with a single `tests_web_app.py` file.

2 - Yep. This parent class should have 3 methods and should be responsible for creating the sketchbook_dir (under `setUp`), clearing all created data by removing the same dir (under `tearDown`) and to set up the flask app (under `create_app`).

File: pyp5js/http/web_app.py (hunk 1 above)
Comment (isabarros) [reply pyp5js/http/web_app.py]: Sure!

File: pyp5js/http/web_app.py [hunk 5]
```diff
@@ -35,16 +35,17 @@ def add_new_sketch_view():
...
-            template = 'new_sketch_success.html'
-            context.update({
-                'files': files,
-                'sketch_url': f'/sketch/{sketch_name}/',
-            })
-        except SketchDirAlreadyExistException:
-            path = SKETCHBOOK_DIR.joinpath(sketch_name)
-            context['error'] = f"The sketch {path} already exists."
+        else:
+            try:
```
Comment (isabarros) [inline pyp5js/http/web_app.py]: @berinhard had to add the `else` because after creating the test for the empty sketch_name scenario it failed because even though it falls in the `if` the try block is executed anyway and the error message is overwritten with `'The sketch {path} already exists.'` message

Comment (berinhard) [reply pyp5js/http/web_app.py]: For sure! Nice catch!

**Comment by isabarros** [comment]
@berinhard Almost done, just one test missing. Please check if the tests expectations are correct.

**Comment by isabarros** [comment]
@berinhard From my perspective it makes more sense to have the scenario `Ensure the sketches are being listed with valid links` as part of SketchFiles class tests than the web_app tests. But would like to hear your thoughts on that.

**Comment by isabarros** [comment]
@berinhard One last important thing to notice is that the coverage might not be very accurate for the flask tests. You can confirm commenting out all the tests in test_web_app file except `test_get_home_renders_index_template` and you will see that the coverage still remains really high (93%). From the article that you sent as reference they use `coverage` lib for that, but I haven't check yet.

File: pyp5js/tests/http/test_web_app.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+            shutil.rmtree(SKETCHBOOK_DIR)
+
+    def create_sketch(self, name):
+        sketch_dir = SKETCHBOOK_DIR.joinpath(name)
+        os.makedirs(sketch_dir)
+        target_file = sketch_dir.joinpath(f'{name}.py')
+        index_file = sketch_dir.joinpath('index.html')
+        self.create_file(target_file)
+        self.create_file(index_file, 'index file content')
+        return sketch_dir
```
Comment (berinhard) [inline pyp5js/tests/http/test_web_app.py]: A refactoring suggestion here is to use the `commands` API. It already handles all the required files and directory creation. So it's highly possible that you can refactor this to:

```python
from pyp5js import commands

def create_sketch(self, name):
    sketch_files = commands.new(name)
    return sketch_files.sketch_dir
```

File: pyp5js/tests/http/test_web_app.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+        sketch_dir = SKETCHBOOK_DIR.joinpath(name)
+        static_dir = sketch_dir.joinpath(static_path.split('/')[0])
+        os.makedirs(static_dir)
+        target_file = sketch_dir.joinpath(f'{name}.py')
+        index_file = sketch_dir.joinpath('index.html')
+        static_file = static_dir.joinpath(static_path.split('/')[1])
+        self.create_file(target_file)
+        self.create_file(index_file)
+        self.create_file(static_file, 'static file content')
+        return sketch_dir
```
Comment (berinhard) [inline pyp5js/tests/http/test_web_app.py]: Same refactoring tip here but now combining the `new` command with the `transcrypt`. The second one is responsible for creating the static files and all the other dependencies.

```python
def create_sketch_with_static_files(self, name):
    sketch_files = commands.new(name)
    commands.transcrypt(name)
    return sketch_fiiles.sketch_dir
```
```

File: pyp5js/tests/http/test_web_app.py [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+        with file_name.open('w') as fd:
+            fd.write(content)
+
+
+class IndexViewTests(Pyp5jsWebTestCase):
+    route = '/'
+
+    def test_get_home_renders_index_template(self):
+        self.client.get(self.route)
+        self.assert_template_used('index.html')
```
Comment (berinhard) [inline pyp5js/tests/http/test_web_app.py]: By ` Ensure the sketches are being listed with valid links` I was thinking on making sure we a valid context with `sketchbook_dir` being the `SKETCHBOOK_DIR` config and with the `sketches` being a list with dicts such as `{"name": "sketch_name", "url": "/sketches/sketch_name"}`.

**Comment by isabarros** [comment]
@berinhard thanks! and sure will do those refactorings! :)

File: pyp5js/tests/http/test_web_app.py (hunk 8 above)
Comment (isabarros) [reply pyp5js/tests/http/test_web_app.py]: @berinhard it makes sense, gonna add a test for that!

**Comment by isabarros** [comment]
@berinhard All tests finished. I think this is ready to go, please check :)

**Comment by berinhard** [comment]
Yaaaaay :1st_place_medal:
